spanner/session.go

```diff
@@ -633,6 +686,69 @@ func (p *sessionPool) recordStat(ctx context.Context, m *stats.Int64Measure, n i
 	recordStat(ctx, m, n)
 }
 
+func (p *sessionPool) getRatioOfSessionsInUseLocked() float64 {
+	maxSessions := p.MaxOpened
+	if maxSessions == 0 {
+		return 0
+	}
+	return float64(p.numInUse) / float64(maxSessions)
+}
+
+// gets sessions which are unexpectedly long-running.
+func (p *sessionPool) getLongRunningSessionsLocked() []*sessionHandle {
+	usedSessionsRatio := p.getRatioOfSessionsInUseLocked()
+	var longRunningSessions []*sessionHandle
+	if usedSessionsRatio > p.usedSessionsRatioThreshold {
+		element := p.trackedSessionHandles.Front()
+		for element != nil {
+			sh := element.Value.(*sessionHandle)
+			sh.mu.Lock()
+			diff := time.Now().Sub(sh.checkoutTime)
+			if !sh.isLongRunningTransaction && diff.Seconds() >= p.idleTimeThreshold.Seconds() {
+				if p.LogInactiveTransactions && !sh.isSessionLeakLogged {
+					if sh.stack != nil {
+						logf(p.sc.logger, "session %s checked out of pool at %s is long running due to possible session leak for goroutine: \n%s", sh.session.getID(), sh.checkoutTime.Format(time.RFC3339), sh.stack)
+					} else {
+						logf(p.sc.logger, "session %s checked out of pool at %s is long running due to possible session leak for goroutine: \nStack trace not available", sh.session.getID(), sh.checkoutTime.Format(time.RFC3339))
+					}
+					sh.isSessionLeakLogged = true
+				}
+				if p.CloseInactiveTransactions && !sh.isSessionLeakLogged {
+					longRunningSessions = append(longRunningSessions, sh)
+					if sh.stack != nil {
+						logf(p.sc.logger, "session %s checked out of pool at %s is long running and will be removed due to possible session leak for goroutine: \n%s", sh.session.getID(), sh.checkoutTime.Format(time.RFC3339), sh.stack)
+					} else {
+						logf(p.sc.logger, "session %s checked out of pool at %s is long running and will be removed due to possible session leak for goroutine: \nStack trace not available", sh.session.getID(), sh.checkoutTime.Format(time.RFC3339))
+					}
+					sh.isSessionLeakLogged = true
```

Comment:
Nit : There is minor duplication in these two blocks. Can we refactor a bit?
* `sh.isSessionLeakLogged = true` is repeated and can be done only once. This can be simplified if we add a check at L707 as 

```
if !sh.isSessionLeakLogged {
     sh.isSessionLeakLogged = true;
}
```

Reply:
Yes, this can be done. But I guess this will come after L723. If we do in L707, then the other two if checks will be ignored always.

But, I wonder refactoring this would add an additional `if` check to make the `isSessionLeakLogged ` property `true`. which could have been done with in the `if` statements that will get executed already. WDYT?